File: bindings/python/src/ptiff/image.py
import contextlib

from .camera import Camera
from .ptiff import (
    ptiff_image_descriptor,
    ptiff_sink_close,
    ptiff_sink_create,
    ptiff_sink_create_camera,
    ptiff_sink_tile_byte_size,
    ptiff_sink_tile_columns,
    ptiff_sink_tile_rows,
    ptiff_sink_write_tile,
    ptiff_source_close,
    ptiff_source_descriptor,
    ptiff_source_open,
    ptiff_source_read_tile,
    ptiff_source_tile_byte_size,
    ptiff_source_tile_columns,
    ptiff_source_tile_rows,
    ptiff_tile_info,
)
from .tile import Tile


class Image:
    """A PTIFF image: metadata plus tile access.

    Use the classmethods `Image.open(path)` (read a TIFF/BigTIFF) and
    `Image.create(path, width, height, pixel_type=...)` (write a new one).
    Reading exposes the structured camera calibration via `img.camera` and the
    raw tile bytes via `img.read_tile(col, row)`.
    """

    def __init__(self, path, handle=None, camera=None):
        self.path = path
        self._handle = handle
        self._sink = None
        self._camera = camera if camera is not None else Camera.from_path(path)
        self._desc = None

    @classmethod
    def open(cls, path):
        path = str(path)
        result = ptiff_source_open(path)
        # SWIG surfaces the optional err_out as a second list element; the
        # handle is always the first.
        handle = result[0] if isinstance(result, (list | tuple)) else result
        if not handle:
            raise OSError(f"ptiff_source_open failed for {path!r}")
        return cls(path, handle=handle)

    @classmethod
    def create(
        cls,
        path,
        width,
        height,
        pixel_type=0,
        tile_width=0,
        tile_height=0,
        channel_count=1,
        camera=None,
    ):
        path = str(path)
        desc = ptiff_image_descriptor()
        desc.width = int(width)
        desc.height = int(height)
        desc.pixel_type = int(pixel_type)
        desc.channel_count = int(channel_count)
        if tile_width > 0 and tile_height > 0:
            ti = ptiff_tile_info()
            ti.tile_width = int(tile_width)
            ti.tile_height = int(tile_height)
            desc.tile_info = ti
            desc.has_tile_info = 1
        if camera is not None:
            cam = camera.to_metadata() if isinstance(camera, Camera) else camera
            sink = ptiff_sink_create_camera(path, desc, cam)
            if not sink:
                raise OSError(f"ptiff_sink_create_camera failed for {path!r}")
        else:
            sink = ptiff_sink_create(path, desc)
            if not sink:
                raise OSError(f"ptiff_sink_create failed for {path!r}")
        obj = cls(path)
        obj._sink = sink
        obj._desc = desc
        obj._camera = camera
        return obj

    # -- metadata ------------------------------------------------------------
    def _descriptor(self):
        if self._desc is None:
            self._desc = ptiff_image_descriptor()
            ptiff_source_descriptor(self._handle, self._desc)
        return self._desc

    @property
    def descriptor(self):
        return self._descriptor()

    @property
    def width(self):
        return self._descriptor().width

    @property
    def height(self):
        return self._descriptor().height

    @property
    def channel_count(self):
        return self._descriptor().channel_count

    @property
    def pixel_type(self):
        return self._descriptor().pixel_type

    @property
    def camera(self):
        return self._camera

    @property
    def tile_columns(self):
        if self._handle:
            return int(ptiff_source_tile_columns(self._handle))
        if self._sink:
            return int(ptiff_sink_tile_columns(self._sink))
        return 0

    @property
    def tile_rows(self):
        if self._handle:
            return int(ptiff_source_tile_rows(self._handle))
        if self._sink:
            return int(ptiff_sink_tile_rows(self._sink))
        return 0

    @property
    def tile_byte_size(self):
        if self._handle:
            return int(ptiff_source_tile_byte_size(self._handle))
        if self._sink:
            return int(ptiff_sink_tile_byte_size(self._sink))
        return 0

    def read_tile(self, column, row):
        if self._handle is None:
            raise OSError("image not open for reading")
        nbytes = self.tile_byte_size
        buf = bytearray(nbytes)
        ptiff_source_read_tile(self._handle, int(column), int(row), buf)
        return Tile(int(column), int(row), bytes(buf), byte_size=nbytes)

    def write_tile(self, column, row, data):
        if self._sink is None:
            raise OSError("image not open for writing")
        buf = bytearray(data)
        ptiff_sink_write_tile(self._sink, int(column), int(row), buf)

    def close(self):
        if self._handle is not None:
            with contextlib.suppress(Exception):
                ptiff_source_close(self._handle)
            self._handle = None
        if self._sink is not None:
            with contextlib.suppress(Exception):
                ptiff_sink_close(self._sink)
            self._sink = None

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()
        return False

    def __repr__(self):
        return f"<ptiff.Image path={self.path!r} {self.width}x{self.height}>"

File: bindings/ruby/lib/ptiff/image.rb
# frozen_string_literal: true

module PTiff
  # Image is a PTIFF image: metadata plus tile access.
  #
  # Use the class methods `Image.open(path)` (read a TIFF/BigTIFF) and
  # `Image.create(path, width, height, pixel_type: ...)` (write a new one).
  # Reading exposes the structured camera calibration via `#camera` and the raw
  # tile bytes via `#read_tile(col, row)`.
  class Image
    attr_reader :path

    def initialize(path, handle: nil, sink: nil, camera: nil, desc: nil)
      @path = path
      @handle = handle
      @sink = sink
      @camera = camera
      @desc = desc
      @camera ||= Camera.from_path(path) if handle || File.file?(path)
    end

    # Opens the TIFF/BigTIFF file at path for reading. Raises IOError on failure.
    def self.open(path)
      path = path.to_s
      res = Ptiff::ptiff_source_open(path)
      handle = res.respond_to?(:first) ? res.first : res
      raise IOError, "ptiff_source_open failed for #{path.inspect}" if handle.nil?
      new(path, handle: handle)
    end

    # Creates a new image at path with the given size and pixel type.
    #
    #   pixel_type:  a PTIFF_PIXEL_* value (default UInt8)
    #   tile_width / tile_height: 0 (the default) disables tiling
    #   channel_count:  defaults to 1
    #   camera:  an optional PTiff::Camera to persist alongside the image
    #
    # Raises IOError if the sink cannot be created.
    def self.create(path, width, height, pixel_type: 0, tile_width: 0,
                    tile_height: 0, channel_count: 1, camera: nil)
      path = path.to_s
      desc = Ptiff::Ptiff_image_descriptor.new
      desc.width = Integer(width)
      desc.height = Integer(height)
      desc.pixel_type = Integer(pixel_type)
      desc.channel_count = Integer(channel_count)
      if tile_width.to_i > 0 && tile_height.to_i > 0
        desc.has_tile_info = 1
        desc.tile_info.tile_width = Integer(tile_width)
        desc.tile_info.tile_height = Integer(tile_height)
      end
      desc.has_compression = 0

      if camera
        struct = camera.is_a?(Camera) ? camera.to_struct : camera
        sink = Ptiff::ptiff_sink_create_camera(path, desc, struct)
        raise IOError, "ptiff_sink_create_camera failed for #{path.inspect}" if sink.nil?
      else
        sink = Ptiff::ptiff_sink_create(path, desc)
        raise IOError, "ptiff_sink_create failed for #{path.inspect}" if sink.nil?
      end
      new(path, sink: sink, desc: desc, camera: camera)
    end

    # Core descriptor as read back from the source (or built on create).
    def descriptor
      @desc ||= begin
        d = Ptiff::Ptiff_image_descriptor.new
        Ptiff::ptiff_source_descriptor(@handle, d)
        d
      end
    end

    def width          = descriptor.width
    def height         = descriptor.height
    def channel_count  = descriptor.channel_count
    def pixel_type     = descriptor.pixel_type
    def camera         = @camera

    def tile_columns
      @handle ? Ptiff::ptiff_source_tile_columns(@handle) : (@sink ? Ptiff::ptiff_sink_tile_columns(@sink) : 0)
    end

    def tile_rows
      @handle ? Ptiff::ptiff_source_tile_rows(@handle) : (@sink ? Ptiff::ptiff_sink_tile_rows(@sink) : 0)
    end

    def tile_byte_size
      @handle ? Ptiff::ptiff_source_tile_byte_size(@handle) : (@sink ? Ptiff::ptiff_sink_tile_byte_size(@sink) : 0)
    end

    # Reads the tile at grid position (column, row) and returns a Tile with the
    # raw byte payload. Requires the image to be open for reading.
    def read_tile(column, row)
      raise IOError, "image not open for reading: #{@path.inspect}" if @handle.nil?
      n = tile_byte_size
      buf = "\0".b * n
      _rc, _nread = Ptiff::ptiff_source_read_tile(@handle, Integer(column), Integer(row), buf)
      Tile.new(Integer(column), Integer(row), buf.dup.force_encoding(Encoding::BINARY))
    end

    # Writes data to the tile at grid position (column, row). Requires the image
    # to be open for writing (created via Image.create).
    def write_tile(column, row, data)
      raise IOError, "image not open for writing: #{@path.inspect}" if @sink.nil?
      buf = String(data).b
      rc = Ptiff::ptiff_sink_write_tile(@sink, Integer(column), Integer(row), buf)
      raise IOError, "ptiff_sink_write_tile(#{column},#{row}) failed (rc=#{rc})" unless rc.zero?
      rc
    end

    # Releases the underlying source/sink handles (a no-op when already closed).
    def close
      unless @handle.nil?
        begin
          Ptiff::ptiff_source_close(@handle)
        rescue StandardError
          nil
        end
        @handle = nil
      end
      unless @sink.nil?
        begin
          Ptiff::ptiff_sink_close(@sink)
        rescue StandardError
          nil
        end
        @sink = nil
      end
    end

    def inspect
      "#<PTiff::Image path=#{@path.inspect} #{width}x#{height}>"
    end
  end
end
